cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -21,6 +21,49 @@ private predicate predictableInstruction(Instruction instr) {
   predictableInstruction(instr.(UnaryInstruction).getUnary())
 }
 
+/**
+ * Functions that we should only allow taint to flow through (to the return
+ * value) if all but the source argument are 'predictable'.  This is done to
+ * emulate the old security library's implementation rather than due to any
+ * strong belief that this is the right approach.
+ *
+ * Note that the list itself is not very principled; it consists of all the
+ * functions listed in the old security library's [default] `isPureFunction`
+ * but not in the old taint tracking library's `returnArgument` predicate.
+ */
+predicate predictableOnlyFlow(string name) {
+  name = "abs" or
+  name = "atof" or
+  name = "atoi" or
+  name = "atol" or
+  name = "atoll" or
+  name = "labs" or
+  name = "strcasestr" or
+  //name = "strcat" or
+  name = "strchnul" or
+  name = "strchr" or
+  name = "strchrnul" or
+  name = "strcmp" or
+  //name = "strcpy" or
+  name = "strcspn" or
+  name = "strdup" or
+  name = "strlen" or
+  //name = "strncat" or
+  name = "strncmp" or
+  //name = "strncpy" or
+  name = "strndup" or
+  name = "strnlen" or
+  name = "strrchr" or
+  name = "strspn" or
+  name = "strstr" or
+  name = "strtod" or
+  name = "strtof" or
+  name = "strtol" or
+  name = "strtoll" or
+  name = "strtoq" or
+  name = "strtoul"
+}
```

Comment:
It's possible the implementation here could be something like `isPureFunction(f.getName()) and not returnArgument(f, _)`, but I'm not keen to import `returnArgument` from the old security library to do this.  Doing so would automatically track any changes to those predicates that are made in future ... but I'm not convinced that's what we want, since this is really nothing more than a weird list of exceptions to get the behaviour we want on cases the old library was tuned for.

I might be more keen if I thought `predictable` was a good idea in the first place.

Reply:
Do the single-argument functions need to be in this list at all?

Reply:
They shouldn't matter.  I should verify that with testing though.

Reply:
Done.  It turns out `strlen` had to remain as well despite being one argument, because it's special case is itself slightly special.